packages/block-editor/src/components/block-alignment-visualizer/index.js

```diff
@@ -0,0 +1,391 @@
+								border-left: solid 2px var(--contrast-color);
+								border-right: solid 2px var(--contrast-color);
+							}
+						` }
+					</style>
+					<LayoutStyle
+						blockName={ blockListBlockName }
+						layout={ layout }
+						selector=".block-editor__alignment-visualizer-zone"
+					/>
+				</head>
+				<body className="editor-styles-wrapper">
+					{ alignments.map( ( alignment ) => (
+						<BlockAlignmentVisualizerZone
+							key={ alignment.name }
+							alignment={ alignment }
+							justification={ layout.justifyContent }
+							color={ contrastColor }
+							isHighlighted={ alignment.name === highlightedZone }
+							addZone={ ( zone ) => zones.current.add( zone ) }
+							removeZone={ ( zone ) =>
+								zones.current.delete( zone )
+							}
+						/>
+					) ) }
+				</body>
+			</Iframe>
+		</Popover>
+	);
+}
+
+function BlockAlignmentVisualizerZone( {
+	alignment,
+	justification,
+	color,
+	isHighlighted,
+	addZone,
+	removeZone,
+} ) {
+	const [ popoverAnchor, setPopoverAnchor ] = useState( null );
+	const { name } = alignment;
+
+	const updateZonesRef = useRefEffect(
+		( node ) => {
+			const zone = {
+				name,
+				node,
+			};
+			addZone( zone );
+			return () => removeZone( zone );
+		},
+		[ name ]
+	);
+
+	const zoneInnerRefs = useMergeRefs( [ updateZonesRef, setPopoverAnchor ] );
+
+	return (
+		<>
+			<div
+				className={ classnames(
+					'block-editor__alignment-visualizer-zone',
+					{
+						[ `is-content-justification-${ justification }` ]:
+							justification,
+					}
+				) }
+			>
+				<div
+					className={ classnames(
+						'block-editor__alignment-visualizer-zone-inner',
+						alignment.className
+					) }
+					ref={ zoneInnerRefs }
+				/>
+			</div>
+			<Popover
+				anchor={ popoverAnchor }
+				className="block-editor__alignment-visualizer-zone-label-popover"
+				placement="top-end"
+				variant="unstyled"
+				flip
+				resize={ false }
+				shift={ false }
+			>
+				<div
+					className={ classnames(
+						'block-editor__alignment-visualizer-zone-label',
+						{ 'is-highlighted': isHighlighted }
+					) }
+					style={ { color } }
+				>
+					{ alignment.label }
+				</div>
+			</Popover>
+		</>
+	);
+}
+
+function Iframe( { children, ...props } ) {
+	const [ iframeDocument, setIframeDocument ] = useState( null );
+
+	const ref = useRefEffect( ( node ) => {
+		function setDocumentIfReady() {
+			const contentDocument = node?.contentDocument;
+			const documentElement = contentDocument?.documentElement;
+			const readyState = contentDocument?.readyState;
+
+			if ( readyState !== 'interactive' && readyState !== 'complete' ) {
+				return;
+			}
+
+			documentElement.removeChild( contentDocument.head );
+			documentElement.removeChild( contentDocument.body );
+			setIframeDocument( documentElement );
+		}
+
+		node.addEventListener( 'load', setDocumentIfReady );
+
+		return () => {
+			node.removeEventListener( 'load', setDocumentIfReady );
+			setIframeDocument( null );
```

Comment:
I think I've come across something similar where the component unmounted but was still setting state in the `useRefEffect`'s cleanup function. This might be a gotcha for the unofficial `useRefEffect` hook since the timing of the cleanup function is probably not what we expect. 🤔